Software/Program/Code/Core/Src/main.c

```diff
@@ -90,18 +147,121 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_TIM2_Init();
+  MX_ADC3_Init();
   /* USER CODE BEGIN 2 */
 
+   HAL_TIM_Base_Start_IT(&htim2);
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
-  while (1)
-  {
+   while (1)
+   {
+      static enum Direction_E Direction = DIRECTION_NOT_AVAILABLE;
+      static bool UpdateOutput = false;
+      static bool ADC_ConversionInProgress = false;
+      static uint16_t Previous_ADCValue = 0u;
+      static float Frequency = 0.0f;
+
+      /*** HANDLE FRQ_OUT GPIO ***/
+      
+
+      /*** HANDLE USER INPUT FOR HIGH-PULSE TIME ***/
+      // Check if button was pressed...
+      if ( ButtonPressed == true )
+      {
+         ButtonPressed = false;
+         HAL_Delay( DEBOUNCE_WAIT_TIME );  // TODO: Handle debounces without a delay... --> Suggestion: Use another timer and a flag in the ISR that prevents immediate setting of flag...
+
+         // Toggle high-pulse time amounts
+         Direction = (enum Direction_E) ( ( (int)Direction + 1 ) % ( (int)DIRECTION_ERROR + 1 ) );
```

Comment:
Safer practice would be to use a `switch-case` statement instead of performing integer arithmetic on enumerated types. This should work, but due to the divide operator, I actually don't think this is any faster than the `switch-case`. The advantage of this is that it is one line of code, but that appears to be it. The disadvantage here is the potential for using an enumerated value that is out-of-range of the enumeration type definition (e.g., if `DIRECTION_ERROR` was _not_ the last enum value) as well as the potentially slower execution compared with a `swtich-case` due to the modulo operation.